clang/test/CAS/print-cc1-cache-key.c

```diff
@@ -0,0 +1,34 @@
+// REQUIRES: shell
+
+// RUN: rm -rf %t && mkdir -p %t
+// RUN: llvm-cas --cas %t/cas --ingest --data %s > %t/casid
+//
+// RUN: %clang -cc1 -triple x86_64-apple-macos11 \
+// RUN:   -fcas-path %t/cas -fcas-fs @%t/casid -fcache-compile-job \
+// RUN:   -Rcompile-job-cache-miss -emit-obj -o %t/output.o %s 2> %t/output.txt
+//
+// RUN: cat %t/output.txt | sed \
+// RUN:   -e "s/^.*miss for '//" \
+// RUN:   -e "s/' .*$//" > %t/cache-key
+//
+// RUN: not c-index-test core -print-cc1-cache-key -cas %t/cas 2>&1 | FileCheck %s -check-prefix=NO_KEY
+// NO_KEY: missing cc1 cache key
+//
+// RUN: not c-index-test core -print-cc1-cache-key -cas %t/cas -key asdf 2>&1 | FileCheck %s -check-prefix=INVALID_KEY
+// INVALID_KEY: invalid cas-id 'asdf'
+//
+// RUN: not c-index-test core -print-cc1-cache-key -cas %t/cas -key @%t/casid 2>&1 | FileCheck %s -check-prefix=NOT_A_KEY
+// NOT_A_KEY: not a valid cache key
+//
+// RUN: c-index-test core -print-cc1-cache-key -cas %t/cas -key @%t/cache-key | FileCheck %s
```

Comment:
I wonder if this really belongs in `llvm-cas` instead, something like:
```
// RUN: llvm-cas -print-node llvm
```
That would make sense if:
- The cache key were built using `NodeBuilder` (does not exist), using `.addField(StringRef, StringRef)` and `.addField(StringRef, ObjectRef)`
- We had a `NodeReader` that could do a pretty print.

I'm not askig you take all that on, but it does seem wrong that we'd have this logic in c-index-test when every cache key will want (basically) the same treatment.

Reply:
I agree that having a way to print nodes is good, but I'm not sure it replaces a clang-specific printer. Some considerations:

* I don't want to print file contents by default, but I do want to print the other string contents by default. They look the same in the current tree. What would a generic node printer do?
* Command-line arguments
  *  They're null-separated, and I need to print them separately. Unless NodeReader knows about that pattern, it will not be a useful printer for these.
  * I'm breaking lines in the output on arguments that start with "-" which I find gives a better presentation for a command-line (e.g. "-x c" instead of "-x\n-c", but I would be willing to give that up for just each argument on its own line if it came down to it. I'm also adding trailing `\` which lets you copy and paste the command-line easily.

If you think we can do all that in a generic way without llvm-cas needing to know what clang is doing, then I'm open to switching to that when it's available.  If not, then I think this is still useful to be clang-specific.

Reply:
I agree with Ben that as we add clang-specific CAS objects it's useful to have domain-specific printers.
Another way I would use this later on is for having sensible printing for the "include-tree" CAS object.

The thing I'm not clear about is whether we should add such stuff on top of `c-index-test` or whether we should introduce a `clang-cas-test` testing utility, WDYT?